### PR TITLE
8283422: Create a new test for JDK-8254790

### PR DESCRIPTION
**Pull request by mrserb**
The [JDK-8254790](https://bugs.openjdk.java.net/browse/JDK-8254790) initially was reported on macOS.
It was found by the "javax/xml/crypto/dsig/GenerationTests.java" test which for some reason worked fine on Linux/Windows most? of the time.

I tried to run that test on Linux/Windows many times on a few systems and it always pass. Also, I got the same result for the tier1...tier4 tests.

I have found that the easiest way to reproduce the crash is to pass the result of the string.indexOf to the str.charAt when the heap is big enough. The test below[1] trigger the crash 1000 time in a row on win/lin on systems I have tested.

I am not sure what is the best way to contribute to the test, should I create a stand-alone test like[1], or update an existing test as I did for the first version of this PR?

[1]:
```
package compiler.intrinsics.string;

import static jdk.test.lib.Asserts.assertTrue;

public final class TestStringIndexOfLargeHeap {
    static final int[] data = new int[Integer.MAX_VALUE / 2];
    public static void main(String[] args) throws Exception {
        test('a');
        test('\ubeef');
    }
    private static void test(char ch) {
        String str = ("0123456789" + ch).repeat(100);
        for (int i = 0; i < 1_000_000; ++i) {
            assertTrue(ch == str.charAt(str.indexOf(ch)));
        }
    }
}
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8283422](https://bugs.openjdk.java.net/browse/JDK-8283422): Create a new test for JDK-8254790

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7883/head:pull/7883` \
`$ git checkout pull/7883`

Update a local copy of the PR: \
`$ git checkout pull/7883` \
`$ git pull https://git.openjdk.java.net/jdk pull/7883/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7883`

View PR using the GUI difftool: \
`$ git pr show -t 7883`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7883.diff">https://git.openjdk.java.net/jdk/pull/7883.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/compiler/intrinsics/string/TestStringIntrinsics2.java [hunk 1]
```diff
@@ -43,6 +43,28 @@
...
+ * @test id=LargeHeap
+ * @bug 8254790
+ * @requires vm.bits == "64" & os.maxMemory > 5G
+ * @modules java.base/jdk.internal.misc
+ * @library /test/lib
+ *
+ * @build sun.hotspot.WhiteBox
+ * @run driver jdk.test.lib.helpers.ClassFileInstaller sun.hotspot.WhiteBox
+ *
+ * @run main/othervm
```
Comment (mrserb) [inline test/hotspot/jtreg/compiler/intrinsics/string/TestStringIntrinsics2.java:56]: Is it useful to run all other tests in the TestStringIntrinsics2 using the large heap?

File: test/hotspot/jtreg/compiler/intrinsics/string/TestStringIntrinsics2.java [hunk 2]
```diff
@@ -43,6 +43,28 @@
  *        compiler.intrinsics.string.TestStringIntrinsics2
  */
 
+/*
+ * @test id=LargeHeap
```
Comment (mrserb) [inline test/hotspot/jtreg/compiler/intrinsics/string/TestStringIntrinsics2.java:47]: Is it useful to run all other tests in the TestStringIntrinsics2 using the large heap?

**Comment by TobiHartmann** [comment]
The test fails in our CI with:

```
java.lang.OutOfMemoryError: Java heap space
	at compiler.intrinsics.string.TestStringIntrinsics2.main(TestStringIntrinsics2.java:142)
	at java.base/java.lang.invoke.LambdaForm$DMH/0x0000000801002800.invokeStatic(LambdaForm$DMH)
	at java.base/java.lang.invoke.LambdaForm$MH/0x0000000801003c00.invoke(LambdaForm$MH)
	at java.base/java.lang.invoke.Invokers$Holder.invokeExact_MT(Invokers$Holder)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invokeImpl(DirectMethodHandleAccessor.java:155)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:577)
	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:127)
	at java.base/java.lang.Thread.run(Thread.java:828)
```

**Comment by TobiHartmann** [comment]
I think the root cause is simply that there is not enough memory available to the VM to allocate such a large object. `MaxRAMPercentage` is set to 4 but the test could also be executed on a system with a small memory size. I think the test needs to be robust enough to account for all kinds of machines/settings.

Do you know why the issue is only reproducible with a large heap and only if the heap is filled?

**Comment by mrserb** [comment]
> I think the root cause is simply that there is not enough memory available to the VM to allocate such a large object. `MaxRAMPercentage` is set to 4 but the test could also be executed on a system with a small memory size. I think the test needs to be robust enough to account for all kinds of machines/settings.

It has a "guard" jtreg tag "os.maxMemory > 5G" which should prevent execution of this test on small systems. I'll bump this requirement in the next revision of the test.

> Do you know why the issue is only reproducible with a large heap and only if the heap is filled?

That's due to the nature of that [bug](https://bugs.openjdk.java.net/browse/JDK-8254790?focusedCommentId=14374611&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14374611) 
> the 32 bit arithmetic instruction used for 64-bit address

**Comment by mrserb** [comment]
I have found the bug where some tests tried to allocate same huge amount of memory and fail:
https://bugs.openjdk.java.net/browse/JDK-8227645
The solution was to move them to the special folder "hotspot/jtreg/resourcehogs" so they will have "exclusiveAccess" flag.
I'll "move" the new test there.

**Comment by mrserb** [comment]
The test is updated:
 - The special wrapper `TestStringIntrinsics2LargeArray` for the `TestStringIntrinsics2` is added to the test/hotspot/jtreg/resourcehogs/. This approoch is based on the similar [test](https://github.com/openjdk/jdk/blob/edb42d7b0aabd07ef9f46cb1f0e862f7bb912378/test/hotspot/jtreg/resourcehogs/serviceability/sa/LingeredAppWithLargeArray.java#L28). Note that the tests in the `resourcehogs` folder use the `exclusiveAccess` property.
 - The new test will be executed as part of the hotspot:tier3

**Review by vnkozlov — COMMENTED** [review]
I agree with your proposed test configuration. Let me test it before approval.

**Review by vnkozlov — APPROVED** [review]
Testing results are good.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me too.

**Pull request closed without merging**